imap_processing/ultra/l2/ultra_l2.py

```diff
@@ -0,0 +1,334 @@
+    skymap.data_1d["corrected_count_rate"] = (
+        skymap.data_1d["counts"] / skymap.data_1d["exposure_factor"]
+    ) - skymap.data_1d["background_rates"]
+
+    # Calculate flux as corrected_counts / (sensitivity * solid_angle * delta_energy)
+    skymap.data_1d["flux"] = skymap.data_1d["corrected_count_rate"] / (
+        skymap.data_1d["sensitivity"] * skymap.solid_angle * delta_energy
+    )
+
+    skymap.data_1d["flux_uncertainty"] = (skymap.data_1d["counts"] ** 0.5) / (
+        skymap.data_1d["exposure_factor"]
+        * skymap.data_1d["sensitivity"]
+        * skymap.solid_angle
+        * delta_energy
+    )
+
+    # Drop the variables that are no longer needed
+    skymap.data_1d = skymap.data_1d.drop_vars(
+        VARIABLES_TO_DROP_AFTER_FLUX_CALCULATION,
+        errors="ignore",
+    )
+
+    return skymap
+
+
+def ultra_l2(
+    data_dict: dict[str, xr.Dataset | str],
+    data_version: str,
+    output_map_properties: MapProperties = DEFAULT_L2_MAP_PROPERTIES,
+) -> list[xr.Dataset]:
+    """
+    Generate and format Ultra L2 ENA Map Product from L1C Products.
+
+    Parameters
+    ----------
+    data_dict : dict[str, xr.Dataset]
+        Dict mapping l1c product identifiers to paths/Datasets containing l1c psets.
+    data_version : str
+        Version of the data product being created.
+    output_map_properties : MapProperties, optional
+        Properties of the map to be generated.
+        Default is defined in `map_properties.py`.
+
+    Returns
+    -------
+    list[xarray.Dataset,]
+        L2 output dataset containing map of the counts on the sky.
+        Wrapped in a list for consistency with other product levels.
+    """
+    l1c_products = data_dict.values()
+    num_l1c_products = len(l1c_products)
+    logger.info(f"Running ultra_l2 processing on {num_l1c_products} L1C products")
+
+    # Regardless of the output sky tiling type, we will directly
+    # project the PSET values into a healpix map. However, if we are outputting
+    # a Healpix map, we can go directly to map with desired nside, nested params
+    healpix_skymap = generate_ultra_healpix_skymap(
+        ultra_l1c_psets=list(l1c_products),
+        output_map_properties=output_map_properties,
+    )
+
+    # Output formatting for HEALPIX tiling
+    if output_map_properties.sky_tiling_type is ena_maps.SkyTilingType.HEALPIX:
+        map_dataset = healpix_skymap.to_dataset()
+        # Add attributes related to the map
+        map_attrs = {
+            "HEALPix_nside": output_map_properties.nside,
+            "HEALPix_nest": output_map_properties.nested,
+            "Data_version": data_version,
+        }
+
+    # TODO: Implement conversion to Rectangular map
+    elif output_map_properties.sky_tiling_type is ena_maps.SkyTilingType.RECTANGULAR:
+        map_attrs = {
+            "Spacing_degrees": output_map_properties.spacing_deg,
+            "Data_version": data_version,
+        }
+        pass
```

Comment:
Raise so we don't think this is working?
```suggestion
        raise NotImplementedError
```

Reply:
Yep!